### PR TITLE
update LSTM chapter

### PR DESCRIPTION
**Pull request by MorganeLu**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] [local build the book](https://github.com/open-academy/machine-learning/blob/main/CONTRIBUTING.md#build-the-book)
- [ ] review the artifacts from local build
- [ ] [local build the slides](https://github.com/open-academy/machine-learning/blob/main/CONTRIBUTING.md#build-the-slides-optional) if appliable
- [ ] Unit test if appliable

# Checklist:

- [ ] My code follows the [style guidelines](https://github.com/open-academy/machine-learning/blob/main/STYLE_GUIDE.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have applied tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md [hunk 1]
```diff
@@ -13,4 +13,635 @@ kernelspec:
...
+
+As we have learned the normal RNNs on the previous lessons, today we are going to use long-short term memory (LSTM) to fix the problems that are induced from the RNN architecture.
+
+```{seealso}
+The vanilla LSTM is proposed in 2005, the paper is [Bidirectional LSTM Networks for Improved Phoneme Classification and Recognition](https://link.springer.com/chapter/10.1007/11550907_126), and after that, a lot of paper based on it appeared.
+```
+
+## Overview
+The essential problems of RNN are vanishing/exploding gradient problems.The gradient vanishing/exploding problem is a common issue in training deep neural networks. This problem occurs when the gradient of the loss function with respect to the weights of the network becomes very small or very large.
+There are several possibilities will bring on this kind of problem, for example:
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/LSTM.md:26]: Please always give empty lines between paragraphs.

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md [hunk 2]
```diff
@@ -13,4 +13,635 @@ kernelspec:
...
+- a poor choice of hyper-parameters
+- a poor architecture
+- a bug in the code
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/LSTM.md:29]: Give comma/period at the end

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md [hunk 3]
```diff
@@ -13,4 +13,635 @@ kernelspec:
...
+
+Luckily, LSTM can use a memory cell for modeling long-range dependencies and avoid vanishing/exploding gradient problems.
+
+First, let's see the architecture of LSTM cell.
+
+:::{figure-md} LSTM cell
+<img src="../../images/deep-learning/LSTM/LSTM_cell.png" width="90%" class="bg-white mb-1">
+
+Illustration of LSTM unit
+:::
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/LSTM.md:39]: PLease always add the reference to the image source.

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md [hunk 4]
```diff
@@ -13,4 +13,635 @@ kernelspec:
...
+
+It looks like the cell of RNN, but there are still some differences between them. The same part is the direction of data, one input and two outputs. 
+However, the LSTM cell adds a cell state $c^{<t>}$, which updates with the time. $c^{<t-1>}$ means cell state at previous time step and $c^{<t-1>}$ means cell state at current time step.
+The $h^{t}$ is still the activation.
+
+:::{figure-md} LSTM cell state
+<img src="../../images/deep-learning/LSTM/cell_state.png" width="90%" class="bg-white mb-1">
+
+LSTM unit of state c
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/LSTM.md:48]: `$c$`. Add image source reference.

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md [hunk 5]
```diff
@@ -13,4 +13,635 @@ kernelspec:
   name: python3
 ---
 
-# LSTM (TBD)
+# Long-Short Term Memory
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/LSTM.md:16]: We only have the first letter as upper case in title.

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md [hunk 6]
```diff
@@ -13,4 +13,635 @@ kernelspec:
...
+
+:::{figure-md} LSTM cell state
+<img src="../../images/deep-learning/LSTM/cell_state.png" width="90%" class="bg-white mb-1">
+
+LSTM unit of state c
+:::
+
+## Special Gates
+Then, we take a look at the inside of cell. $\bigodot$ means element-wise multiplication operator, $\bigoplus$ means element-wise addition operator and $\sigma$ means logistic sigmoid activation functions.
+Now, we should pay attention to 'Gates'.
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/LSTM.md:53]: empty lines.

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md [hunk 7]
```diff
@@ -13,4 +13,635 @@ kernelspec:
...
+:::
+
+## Special Gates
+Then, we take a look at the inside of cell. $\bigodot$ means element-wise multiplication operator, $\bigoplus$ means element-wise addition operator and $\sigma$ means logistic sigmoid activation functions.
+Now, we should pay attention to 'Gates'. 
+
+The first is 'Forget Gate' $f$, and it controls which information is remembered, and which is forgotten; can reset the cell state. This function can be written as $f_t = \sigma (W_{fx}x^{<t>} + W_{fh}h^{<t-1>} + b_f)$
+
+Next, 'Input Gate' is $i_t = \sigma(W_{ix}x^{<t>} + W_{ih}h^{<t-1>} + b_i)$ and 'Input Node' is $g_t = tanh(W_{gt}x^{<t>} + W_{gh}x^{<t-1>} + b_g)$.
+To brief summarize the previous gates in an expression: $C^{<t>} = (C^{<t-1>} \bigodot f_t) \bigoplus (i_t \bigodot g_t)$. Since $i_t$ is 'Input Gate' and $g_t$ is 'Input Gate', $(i_t \bigodot g_t)$ is for updatting the cell state.
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/LSTM.md:58]: empty line

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md [hunk 8]
```diff
@@ -13,4 +13,635 @@ kernelspec:
...
+
+Practice the Long-Short Term Memory Networks by following this TBD.
+
+## Self study
+
+TBD
+
+## Acknowledgments
+
+Thanks to [Nick](https://github.com/nfmcclure) for creating the open-source course [tensorflow_cookbook](https://github.com/nfmcclure/tensorflow_cookbook) and [Sebastian Raschka](https://github.com/rasbt) for creating the open-sourse [stat453-deep-learning-ss20](https://github.com/rasbt/stat453-deep-learning-ss20). It inspires the majority of the content in this chapter.
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/LSTM.md:641]: It inspires -> They inspire

Comment (zhangqi444) [reply open-machine-learning-jupyter-book/deep-learning/LSTM.md:641]: open-sourse -> open-source

Comment (zhangqi444) [reply open-machine-learning-jupyter-book/deep-learning/LSTM.md:641]: is the tensorflow_cookbook a course? is the stat453-deep-learning-ss20 a course. Please correct the statement.

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md [hunk 9]
```diff
@@ -13,4 +13,635 @@ kernelspec:
...
+plt.show()
+```
+
+## Your turn! 🚀
+
+Practice the Long-Short Term Memory Networks by following this TBD.
+
+## Self study
+
+TBD
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/LSTM.md:637]: remove if you don't have such content.

**Review by zhangqi444 — CHANGES_REQUESTED** [review]

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md (hunk 1 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/LSTM.md:27]: Done.

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md (hunk 2 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/LSTM.md]: Done.

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md (hunk 3 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/LSTM.md]: Done. I have added the reference to the reference.bib.

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md (hunk 4 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/LSTM.md]: Done. I have added the reference to the reference.bib.

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md (hunk 5 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/LSTM.md]: Done.

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md (hunk 6 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/LSTM.md:55]: Done.

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md (hunk 7 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/LSTM.md:61]: Done.

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md (hunk 8 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/LSTM.md]: Done. "Course" has been changed to "library".

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md (hunk 9 above)
Comment (MorganeLu) [reply open-machine-learning-jupyter-book/deep-learning/LSTM.md]: Done.

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md [hunk 10]
```diff
@@ -13,4 +13,721 @@ kernelspec:
...
+plt.show()
+```
+
+## Your turn! 🚀
+
+Practice the Long-Short Term Memory Networks by following this TBD.
+
+## Acknowledgments
+
+Thanks to [Nick](https://github.com/nfmcclure) for creating the open-source library [tensorflow_cookbook](https://github.com/nfmcclure/tensorflow_cookbook) and [Sebastian Raschka](https://github.com/rasbt) for creating the open-source library [stat453-deep-learning-ss20](https://github.com/rasbt/stat453-deep-learning-ss20). They inspire the majority of the content in this chapter.
```
Comment (zhangqi444) [inline open-machine-learning-jupyter-book/deep-learning/LSTM.md:727]: You can just call these as open-source project, which is more general .

**Review by zhangqi444 — APPROVED** [review]

File: open-machine-learning-jupyter-book/deep-learning/LSTM.md (hunk 10 above)
Comment (reveurmichael) [reply open-machine-learning-jupyter-book/deep-learning/LSTM.md:727]: Done

**Pull request merged**